staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff_test.go

```diff
@@ -0,0 +1,508 @@
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	az := GetTestCloud(ctrl)
+	az.nsgCache.Set("sg", "test")
+
+	mockSGClient := az.SecurityGroupsClient.(*mocksecuritygroupclient.MockInterface)
+	mockSGClient.EXPECT().CreateOrUpdate(gomock.Any(), az.ResourceGroup, gomock.Any(), gomock.Any(), gomock.Any()).Return(&retry.Error{
+		RawError: fmt.Errorf(operationCanceledErrorMessage),
+	})
+	mockSGClient.EXPECT().Get(gomock.Any(), az.ResourceGroup, "sg", gomock.Any()).Return(network.SecurityGroup{}, nil)
+
+	err := az.CreateOrUpdateSecurityGroup(&v1.Service{}, network.SecurityGroup{Name: to.StringPtr("sg")})
+	assert.Equal(t, fmt.Errorf("Retriable: false, RetryAfter: 0s, HTTPStatusCode: 0, RawError: canceledandsupersededduetoanotheroperation"), err)
+
+	// security group should be removed from cache if the operation is canceled
+	shouldBeEmpty, err := az.nsgCache.Get("sg", cache.CacheReadTypeDefault)
+	assert.NoError(t, err)
+	assert.Empty(t, shouldBeEmpty)
+}
+
+func TestCreateOrUpdateLB(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	tests := []struct {
+		clientErr   *retry.Error
+		expectedErr error
+	}{
+		{
+			clientErr:   &retry.Error{HTTPStatusCode: http.StatusPreconditionFailed},
+			expectedErr: fmt.Errorf("Retriable: false, RetryAfter: 0s, HTTPStatusCode: 412, RawError: <nil>"),
+		},
+		{
+			clientErr:   &retry.Error{RawError: fmt.Errorf(operationCanceledErrorMessage)},
+			expectedErr: fmt.Errorf("Retriable: false, RetryAfter: 0s, HTTPStatusCode: 0, RawError: canceledandsupersededduetoanotheroperation"),
+		},
+	}
+
+	for _, test := range tests {
+		az := GetTestCloud(ctrl)
+		az.lbCache.Set("lb", "test")
+
+		mockLBClient := az.LoadBalancerClient.(*mockloadbalancerclient.MockInterface)
+		mockLBClient.EXPECT().CreateOrUpdate(gomock.Any(), az.ResourceGroup, gomock.Any(), gomock.Any(), gomock.Any()).Return(test.clientErr)
+		mockLBClient.EXPECT().Get(gomock.Any(), az.ResourceGroup, "lb", gomock.Any()).Return(network.LoadBalancer{}, nil)
+
+		err := az.CreateOrUpdateLB(&v1.Service{}, network.LoadBalancer{
+			Name: to.StringPtr("lb"),
+			Etag: to.StringPtr("etag"),
+		})
+		assert.Equal(t, test.expectedErr, err)
+
+		// loadbalancer should be removed from cache if the etag is mismatch or the operation is canceled
+		shouldBeEmpty, err := az.lbCache.Get("lb", cache.CacheReadTypeDefault)
+		assert.NoError(t, err)
+		assert.Empty(t, shouldBeEmpty)
+	}
+}
+
+func TestListLB(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	az := GetTestCloud(ctrl)
+	mockLBClient := az.LoadBalancerClient.(*mockloadbalancerclient.MockInterface)
+	mockLBClient.EXPECT().List(gomock.Any(), az.ResourceGroup).Return(nil, &retry.Error{HTTPStatusCode: http.StatusInternalServerError})
+
+	pips, err := az.ListLB(&v1.Service{})
+	assert.Equal(t, fmt.Errorf("Retriable: false, RetryAfter: 0s, HTTPStatusCode: 500, RawError: <nil>"), err)
+	assert.Empty(t, pips)
+}
+
+func TestListPIP(t *testing.T) {
```

Comment:
could we add both positive and negative cases for PIP operations? (same for the below)

Reply:
The positive cases have been covered somewhere else.